components/util/src/datap/mod.rs

```diff
@@ -0,0 +1,190 @@
+
+impl From<TypeId> for PayloadError {
+    fn from(type_id: TypeId) -> Self {
+        PayloadError::TypeMismatchError(type_id)
+    }
+}
+
+// TODO: Should this be an implemention of std::borrow::Borrow?
+// TODO: Should the error types be &dyn Any, like for Box<dyn Any>::downcast?
+impl<'d> Response<'d> {
+    /// Get an immutable reference to the payload in a Response object.
+    /// The payload may or may not be owned by the Response.
+    pub fn borrow_payload<T: 'static>(&self) -> Result<&T, PayloadError> {
+        let borrowed: &dyn CloneableAny = self.payload.borrow();
+        borrowed
+            .as_any()
+            .downcast_ref::<T>()
+            .ok_or_else(|| PayloadError::from(borrowed.as_any().type_id()))
+    }
+
+    /// Get a mutable reference to the payload in a Response object.
+    /// The payload may or may not be owned by the Response.
+    pub fn borrow_payload_mut<T: 'static>(&mut self) -> Result<&mut T, PayloadError> {
+        let boxed: &mut Box<dyn CloneableAny> = self.payload.to_mut();
+        let borrowed_mut: &mut dyn CloneableAny = boxed.borrow_mut();
+        // TODO: If I move this into the lambda, I get E0502. Why?
+        let type_id = borrowed_mut.as_any().type_id();
+        borrowed_mut
+            .as_any_mut()
+            .downcast_mut::<T>()
+            .ok_or_else(|| PayloadError::from(type_id))
+    }
+
+    /// Take ownership of the payload from a Response object. Consumes the Response object.
+    pub fn take_payload<T: 'static + Clone>(self) -> Result<Cow<'d, T>, PayloadError> {
+        match self.payload {
+            Cow::Borrowed(borrowed) => match borrowed.as_any().downcast_ref::<T>() {
+                Some(v) => Ok(Cow::Borrowed(v)),
+                None => Err(PayloadError::from(borrowed.as_any().type_id())),
+            },
+            Cow::Owned(boxed) => match boxed.into_any().downcast::<T>() {
+                Ok(boxed_t) => Ok(Cow::Owned(*boxed_t)),
+                Err(boxed_any) => Err(PayloadError::from(boxed_any.type_id())),
+            },
+        }
+    }
+}
+
+/// Builder class used to construct a Response.
+pub struct ResponseBuilder {
+    pub data_locale: String,
+}
+
+impl ResponseBuilder {
+    /// Construct a Response from the builder, with owned data.
+    /// Consumes both the builder and the data.
+    /// Returns the 'static lifetime since there is no borrowed data.
+    pub fn with_owned_payload<T: 'static + Clone + Debug>(self, t: T) -> Response<'static> {
+        Response {
+            data_locale: self.data_locale,
+            payload: Cow::Owned(Box::new(t) as Box<dyn CloneableAny>),
+        }
+    }
+
+    /// Construct a Response from the builder, with borrowed data.
+    /// Consumes the builder, but not the data.
+    pub fn with_borrowed_payload<'d, T: 'static + Clone + Debug>(self, t: &'d T) -> Response<'d> {
+        Response {
+            data_locale: self.data_locale,
+            payload: Cow::Borrowed(t),
+        }
+    }
+}
```

Comment:
Why not put those two methods as constructors on `Response`?

Reply:
Right now I have

```rust
let builder = ResponseBuilder { ... };
let response = builder::with_owned_data(data);
```

I think you're suggesting something more like

```rust
let builder = ResponseBuilder { ... };
let response = Response::from_owned(builder, data);
```

I don't have a strong opinion.  I think it'll be decided based on #112.

Do you mind checking this in as-is, and considering a follow-up PR based on what we decide to be the recommended style in ICU4X?

Reply:
I thought about:

```rust

impl Response {
    ...
    pub fn with_borrowed_payload<'d, T: 'static + Clone + Debug>(locale: String, t: &'d T) -> Self<'d> {
        Self {
            data_locale: locale,
            payload: Cow::Borrowed(t),
        }
    }
}

fn main() {
    let response = Response::with_borrowed_payload(
        "und".to_string(),
        &self.data.decimal.symbols_v1_a
    );
}
```

just treat it as a constructor. What's the value of `ResponseBuilder` struct?

Reply:
> Do you mind checking this in as-is, and considering a follow-up PR based on what we decide to be the recommended style in ICU4X?

Yeah, sure! As I said, you have my r+ and all the suggestions are open ended :)

Reply:
I see.  Yeah, there is only one field in Response, but I anticipate having more.  Let's resolve #112 and revisit this if necessary.